blueflood-core/src/main/java/com/rackspacecloud/blueflood/inputs/processors/TokenDiscoveryWriter.java

```diff
@@ -0,0 +1,149 @@
+package com.rackspacecloud.blueflood.inputs.processors;
+
+import com.codahale.metrics.Meter;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.rackspacecloud.blueflood.cache.LocatorCache;
+import com.rackspacecloud.blueflood.cache.TokenCache;
+import com.rackspacecloud.blueflood.concurrent.FunctionWithThreadPool;
+import com.rackspacecloud.blueflood.io.TokenDiscoveryIO;
+import com.rackspacecloud.blueflood.service.Configuration;
+import com.rackspacecloud.blueflood.service.CoreConfig;
+import com.rackspacecloud.blueflood.types.IMetric;
+import com.rackspacecloud.blueflood.types.Locator;
+import com.rackspacecloud.blueflood.types.Token;
+import com.rackspacecloud.blueflood.utils.Metrics;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * A separate discovery writer for tokens. This class is modelled after {@link DiscoveryWriter}
+ */
+public class TokenDiscoveryWriter extends FunctionWithThreadPool<List<List<IMetric>>, Void> {
+
+    private final List<TokenDiscoveryIO> tokenDiscoveryIOs = new ArrayList<>();
+    private final Map<Class<? extends TokenDiscoveryIO>, Meter> writeErrorMeters = new HashMap<>();
+    private static final Logger log = LoggerFactory.getLogger(TokenDiscoveryWriter.class);
+
+    public TokenDiscoveryWriter(ThreadPoolExecutor executor) {
+        super(executor);
+        registerIOModules();
+    }
+
+    public void registerIO(TokenDiscoveryIO io) {
+        tokenDiscoveryIOs.add(io);
+        writeErrorMeters.put(io.getClass(),
+                             Metrics.meter(io.getClass(), "TokenDiscoveryWriter Write Errors")
+        );
+    }
+
+    public void registerIOModules() {
+        List<String> modules = Configuration.getInstance().getListProperty(CoreConfig.TOKEN_DISCOVERY_MODULES);
+
+        ClassLoader classLoader = TokenDiscoveryIO.class.getClassLoader();
```

Comment:
I just answered above on why we do this. I understand it is making it complex but looks like it is the way it is designed. Provides some form for plug-ability if someone decides to use something other than elastic search.